MongoDB connection errors

Sometimes, the REST response is:
```
{
  “message”: “EOF”,
  “reason”: “ReasonInternal”
 }
```

**How to reproduce**
Stop MongoDB.  

This can **NOT** be fixed after start MongoDB as the session is not refreshed, have to restart Cyclone server.

Reopen this issue as `Closed explicitly` still exists.

Error log from MongoDB:
```
2017-11-27T09:01:47.899+0000 I NETWORK  [conn3] end connection 192.168.65.54:51472 (2 connections now open)
2017-11-27T09:01:47.900+0000 I NETWORK  [conn1] end connection 192.168.65.54:47646 (2 connections now open)
2017-11-27T09:01:47.899+0000 I NETWORK  [conn2] end connection 192.168.65.54:49746 (2 connections now open)
2017-11-27T09:01:52.383+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.23:35430 #4 (1 connection now open)
2017-11-27T09:02:07.384+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.23:35506 #5 (2 connections now open)
2017-11-27T09:05:07.406+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.23:36168 #6 (3 connections now open)
2017-11-27T09:05:08.820+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.23:36170 #7 (4 connections now open)
2017-11-27T09:13:37.945+0000 I COMMAND  [conn4] query cyclone.VersionCollection query: { _id: "5a1bd72e4ba1770001f60bd1" } planSummary: IDHACK ntoskip:0 nscanned:1 nscannedObjects:1 idhack:1 keyUpdates:0 writeConflicts:0 numYields:1 nreturned:1 reslen:235 locks:{ Global: { acquireCount: { r: 4 } }, MMAPV1Journal: { acquireCount: { r: 2 } }, Database: { acquireCount: { r: 2 } }, Collection: { acquireCount: { R: 2 } } } 361ms
2017-11-27T09:14:02.834+0000 I COMMAND  [conn4] query cyclone.VersionCollection query: { _id: "5a1bd72e4ba1770001f60bd1" } planSummary: IDHACK ntoskip:0 nscanned:1 nscannedObjects:1 idhack:1 keyUpdates:0 writeConflicts:0 numYields:0 nreturned:1 reslen:235 locks:{ Global: { acquireCount: { r: 2 } }, MMAPV1Journal: { acquireCount: { r: 1 } }, Database: { acquireCount: { r: 1 } }, Collection: { acquireCount: { R: 1 } } } 611ms
2017-11-27T09:14:44.653+0000 I COMMAND  [conn4] query cyclone.VersionCollection query: { _id: "5a1bd72e4ba1770001f60bd1" } planSummary: IDHACK ntoskip:0 nscanned:1 nscannedObjects:1 idhack:1 keyUpdates:0 writeConflicts:0 numYields:0 nreturned:1 reslen:235 locks:{ Global: { acquireCount: { r: 2 } }, MMAPV1Journal: { acquireCount: { r: 1 } }, Database: { acquireCount: { r: 1 } }, Collection: { acquireCount: { R: 1 } } } 1611ms
2017-11-27T09:15:09.116+0000 I COMMAND  [conn4] query cyclone.VersionCollection query: { _id: "5a1bd72e4ba1770001f60bd1" } planSummary: IDHACK ntoskip:0 nscanned:1 nscannedObjects:1 idhack:1 keyUpdates:0 writeConflicts:0 numYields:1 nreturned:1 reslen:235 locks:{ Global: { acquireCount: { r: 4 } }, MMAPV1Journal: { acquireCount: { r: 2 } }, Database: { acquireCount: { r: 2 } }, Collection: { acquireCount: { R: 2 } } } 968ms
2017-11-27T09:16:17.323+0000 I COMMAND  [conn5] command cyclone.$cmd command: findAndModify { findAndModify: "VersionCollection", query: { _id: "5a1bd7d84ba1770001f60bd2" }, update: { $set: { _id: "5a1bd7d84ba1770001f60bd2", service_id: "5a1bbc11832c780001a1bf6d", ref: "branch2", create_time: new Date(1511774168995), status: "running", operation: "codeCheckout,package,imageBuild,publish,unitTest" } } } update: { $set: { _id: "5a1bd7d84ba1770001f60bd2", service_id: "5a1bbc11832c780001a1bf6d", ref: "branch2", create_time: new Date(1511774168995), status: "running", operation: "codeCheckout,package,imageBuild,publish,unitTest" } } nscanned:1 nscannedObjects:1 nMatched:1 nModified:1 fastmod:1 keyUpdates:0 writeConflicts:0 numYields:0 reslen:293 locks:{ Global: { acquireCount: { r: 1, w: 1 } }, MMAPV1Journal: { acquireCount: { w: 2 } }, Database: { acquireCount: { w: 1 } }, Collection: { acquireCount: { W: 1 } } } 400ms
2017-11-27T13:03:24.871+0000 I NETWORK  [conn6] end connection 192.168.66.23:36168 (3 connections now open)
2017-11-27T13:03:24.872+0000 I NETWORK  [conn5] end connection 192.168.66.23:35506 (2 connections now open)
2017-11-27T13:03:24.877+0000 I NETWORK  [conn7] end connection 192.168.66.23:36170 (1 connection now open)
2017-11-27T13:03:24.878+0000 I NETWORK  [conn4] end connection 192.168.66.23:35430 (0 connections now open)
2017-11-27T13:03:34.595+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.61:52998 #8 (1 connection now open)
2017-11-27T13:04:04.593+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.61:53184 #9 (2 connections now open)
2017-11-27T13:30:43.519+0000 I COMMAND  [conn9] query cyclone.VersionLogCollection query: { version_id: "5a1c130254217b0001e7bc14" } planSummary: COLLSCAN ntoskip:0 nscanned:0 nscannedObjects:22 keyUpdates:0 writeConflicts:0 numYields:5 nreturned:1 reslen:2310 locks:{ Global: { acquireCount: { r: 12 } }, MMAPV1Journal: { acquireCount: { r: 6 } }, Database: { acquireCount: { r: 6 } }, Collection: { acquireCount: { R: 6 } } } 133ms
2017-11-27T20:07:46.489+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.61:54394 #10 (3 connections now open)
2017-11-28T00:09:17.758+0000 I NETWORK  [conn9] end connection 192.168.64.61:53184 (2 connections now open)
2017-11-28T00:09:18.583+0000 I COMMAND  [conn10] command admin.$cmd command: ping { ping: 1 } keyUpdates:0 writeConflicts:0 numYields:0 reslen:37 locks:{} 14897ms
2017-11-28T00:09:18.584+0000 I NETWORK  [conn10] end connection 192.168.64.61:54394 (1 connection now open)
2017-11-28T00:09:24.111+0000 I NETWORK  [conn8] end connection 192.168.64.61:52998 (0 connections now open)
2017-11-28T00:09:36.124+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.61:43526 #11 (1 connection now open)
2017-11-28T01:19:28.856+0000 I COMMAND  [PeriodicTaskRunner] task: DBConnectionPool-cleaner took: 103ms
2017-11-28T01:21:58.725+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.61:57134 #12 (2 connections now open)
2017-11-28T01:21:58.911+0000 I NETWORK  [conn11] end connection 192.168.64.61:43526 (1 connection now open)
2017-11-28T01:21:58.929+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.61:57140 #13 (2 connections now open)
2017-11-28T01:21:58.934+0000 I NETWORK  [conn12] end connection 192.168.64.61:57134 (1 connection now open)
2017-11-28T01:42:11.811+0000 I NETWORK  [conn13] end connection 192.168.64.61:57140 (0 connections now open)
2017-11-28T01:42:15.814+0000 I NETWORK  [initandlisten] connection accepted from 192.168.65.81:38270 #14 (1 connection now open)
2017-11-28T01:43:45.837+0000 I NETWORK  [initandlisten] connection accepted from 192.168.65.81:38422 #15 (2 connections now open)
2017-11-28T01:46:51.247+0000 I COMMAND  [conn14] query cyclone.VersionCollection query: { $query: { service_id: "5a1bbc11832c780001a1bf6d" }, $orderby: { create_time: -1 } } planSummary: COLLSCAN, COLLSCAN ntoreturn:20 ntoskip:0 nscanned:0 nscannedObjects:12 scanAndOrder:1 keyUpdates:0 writeConflicts:0 numYields:0 nreturned:6 reslen:2169 locks:{ Global: { acquireCount: { r: 2 } }, MMAPV1Journal: { acquireCount: { r: 1 } }, Database: { acquireCount: { r: 1 } }, Collection: { acquireCount: { R: 1 } } } 138ms
2017-11-28T02:22:08.344+0000 I COMMAND  [conn14] query cyclone.VersionCollection query: { _id: "5a1cc80d8ddb550001ef881e" } planSummary: IDHACK ntoskip:0 nscanned:1 nscannedObjects:1 idhack:1 keyUpdates:0 writeConflicts:0 numYields:0 nreturned:1 reslen:222 locks:{ Global: { acquireCount: { r: 2 } }, MMAPV1Journal: { acquireCount: { r: 1 } }, Database: { acquireCount: { r: 1 } }, Collection: { acquireCount: { R: 1 } } } 107ms
2017-11-28T02:26:43.396+0000 I COMMAND  [conn14] command cyclone.$cmd command: count { count: "projects", query: { name: "qme-test_w1" } } planSummary: COLLSCAN keyUpdates:0 writeConflicts:0 numYields:2 reslen:44 locks:{ Global: { acquireCount: { r: 6 } }, MMAPV1Journal: { acquireCount: { r: 3 } }, Database: { acquireCount: { r: 3 } }, Collection: { acquireCount: { R: 3 } } } 1029ms
2017-11-28T02:26:47.753+0000 I COMMAND  [conn14] query cyclone.projects query: { name: "qme-test_w1" } planSummary: COLLSCAN ntoskip:0 nscanned:0 nscannedObjects:1 keyUpdates:0 writeConflicts:0 numYields:2 nreturned:1 reslen:279 locks:{ Global: { acquireCount: { r: 6 } }, MMAPV1Journal: { acquireCount: { r: 3 } }, Database: { acquireCount: { r: 3 } }, Collection: { acquireCount: { R: 3 } } } 2551ms
2017-11-28T02:26:48.659+0000 I COMMAND  [conn14] command cyclone.$cmd command: count { count: "pipelines", query: { projectID: "5a1bbbdd832c780001a1bf6c" } } planSummary: COLLSCAN keyUpdates:0 writeConflicts:0 numYields:2 reslen:44 locks:{ Global: { acquireCount: { r: 6 } }, MMAPV1Journal: { acquireCount: { r: 3 } }, Database: { acquireCount: { r: 3 } }, Collection: { acquireCount: { R: 3 } } } 150ms
2017-11-28T02:27:12.211+0000 I COMMAND  [conn14] command cyclone.$cmd command: count { count: "projects", query: { name: "qme-test_w1" } } planSummary: COLLSCAN keyUpdates:0 writeConflicts:0 numYields:1 reslen:44 locks:{ Global: { acquireCount: { r: 4 } }, MMAPV1Journal: { acquireCount: { r: 2 } }, Database: { acquireCount: { r: 2 } }, Collection: { acquireCount: { R: 2 } } } 18396ms
2017-11-28T02:40:54.260+0000 I NETWORK  [conn15] end connection 192.168.65.81:38422 (1 connection now open)
2017-11-28T02:40:54.260+0000 I NETWORK  [conn14] end connection 192.168.65.81:38270 (1 connection now open)
2017-11-28T02:41:01.098+0000 I NETWORK  [initandlisten] connection accepted from 192.168.67.52:32984 #16 (1 connection now open)
2017-11-28T02:41:16.102+0000 I NETWORK  [initandlisten] connection accepted from 192.168.67.52:33066 #17 (2 connections now open)
2017-11-28T02:46:03.728+0000 I NETWORK  [conn16] end connection 192.168.67.52:32984 (1 connection now open)
2017-11-28T02:46:03.729+0000 I NETWORK  [conn17] end connection 192.168.67.52:33066 (1 connection now open)
2017-11-28T02:46:05.938+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.69:39544 #18 (1 connection now open)
2017-11-28T02:46:20.939+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.69:39622 #19 (2 connections now open)
2017-11-28T04:52:03.534+0000 I COMMAND  [conn19] query cyclone.VersionLogCollection query: { version_id: "5a1ccf4c080fa40001e0ccc3" } planSummary: COLLSCAN ntoskip:0 nscanned:0 nscannedObjects:35 keyUpdates:0 writeConflicts:0 numYields:27 nreturned:0 reslen:20 locks:{ Global: { acquireCount: { r: 56 } }, MMAPV1Journal: { acquireCount: { r: 28 } }, Database: { acquireCount: { r: 28 } }, Collection: { acquireCount: { R: 28 } } } 131ms
2017-11-28T06:51:05.656+0000 I STORAGE  [DataFileSync] flushing mmaps took 42115ms  for 4 files
2017-11-28T07:11:17.551+0000 I NETWORK  [initandlisten] connection accepted from 192.168.64.69:54802 #20 (3 connections now open)
2017-11-28T07:50:42.865+0000 I NETWORK  [conn19] end connection 192.168.64.69:39622 (2 connections now open)
2017-11-28T07:50:42.866+0000 I NETWORK  [conn18] end connection 192.168.64.69:39544 (2 connections now open)
2017-11-28T07:50:42.885+0000 I NETWORK  [conn20] end connection 192.168.64.69:54802 (2 connections now open)
2017-11-28T07:50:55.179+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:42486 #21 (1 connection now open)
2017-11-28T07:51:10.177+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:42522 #22 (2 connections now open)
2017-11-28T08:11:55.393+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:45364 #23 (3 connections now open)
2017-11-28T08:16:57.224+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:46062 #24 (4 connections now open)
2017-11-28T08:16:57.512+0000 I NETWORK  [conn23] end connection 192.168.66.36:45364 (3 connections now open)
2017-11-28T08:16:57.616+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:46072 #25 (4 connections now open)
2017-11-28T08:58:10.565+0000 I COMMAND  [conn22] command cyclone.$cmd command: count { count: "VersionCollection", query: { service_id: "5a1d1c11b7be1a0001386065" } } planSummary: COLLSCAN keyUpdates:0 writeConflicts:0 numYields:1 reslen:44 locks:{ Global: { acquireCount: { r: 4 } }, MMAPV1Journal: { acquireCount: { r: 2 } }, Database: { acquireCount: { r: 2 } }, Collection: { acquireCount: { R: 2 } } } 478ms
2017-11-28T09:34:51.417+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56822 #26 (5 connections now open)
2017-11-28T09:34:51.445+0000 I NETWORK  [conn22] end connection 192.168.66.36:42522 (4 connections now open)
2017-11-28T09:34:51.455+0000 I NETWORK  [conn24] end connection 192.168.66.36:46062 (3 connections now open)
2017-11-28T09:34:51.456+0000 I NETWORK  [conn21] end connection 192.168.66.36:42486 (2 connections now open)
2017-11-28T09:34:51.459+0000 I NETWORK  [conn25] end connection 192.168.66.36:46072 (1 connection now open)
2017-11-28T09:34:51.640+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56836 #27 (2 connections now open)
2017-11-28T09:34:51.641+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56848 #28 (3 connections now open)
2017-11-28T09:34:51.641+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56880 #29 (4 connections now open)
2017-11-28T09:34:51.647+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56900 #30 (5 connections now open)
2017-11-28T09:34:51.647+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56916 #31 (6 connections now open)
2017-11-28T09:34:51.647+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56930 #32 (7 connections now open)
2017-11-28T09:34:51.648+0000 I NETWORK  [conn31] end connection 192.168.66.36:56916 (6 connections now open)
2017-11-28T09:34:51.661+0000 I NETWORK  [conn30] SocketException handling request, closing client connection: 9001 socket exception [SEND_ERROR] server [192.168.66.36:56900]
2017-11-28T09:34:51.764+0000 I NETWORK  [conn27] end connection 192.168.66.36:56836 (4 connections now open)
2017-11-28T09:34:51.807+0000 I NETWORK  [conn29] SocketException handling request, closing client connection: 9001 socket exception [SEND_ERROR] server [192.168.66.36:56880]
2017-11-28T09:34:51.940+0000 I NETWORK  [conn26] end connection 192.168.66.36:56822 (2 connections now open)
2017-11-28T09:34:51.941+0000 I NETWORK  [conn28] end connection 192.168.66.36:56848 (1 connection now open)
2017-11-28T09:35:04.015+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:56980 #33 (2 connections now open)
2017-11-28T10:42:17.153+0000 I STORAGE  [DataFileSync] flushing mmaps took 25746ms  for 4 files
2017-11-28T10:42:17.153+0000 I NETWORK  [conn33] end connection 192.168.66.36:56980 (1 connection now open)
2017-11-28T10:42:17.279+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37764 #34 (2 connections now open)
2017-11-28T10:42:17.393+0000 I NETWORK  [conn32] end connection 192.168.66.36:56930 (1 connection now open)
2017-11-28T10:42:17.857+0000 I COMMAND  [PeriodicTaskRunner] task: DBConnectionPool-cleaner took: 26455ms
2017-11-28T10:42:17.857+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37766 #35 (2 connections now open)
2017-11-28T10:42:17.857+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37782 #36 (3 connections now open)
2017-11-28T10:42:17.864+0000 I NETWORK  [conn34] end connection 192.168.66.36:37764 (2 connections now open)
2017-11-28T10:42:17.864+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37784 #37 (4 connections now open)
2017-11-28T10:42:17.864+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37800 #38 (5 connections now open)
2017-11-28T10:42:17.864+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37804 #39 (5 connections now open)
2017-11-28T10:42:17.864+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37818 #40 (6 connections now open)
2017-11-28T10:42:17.870+0000 I NETWORK  [conn37] end connection 192.168.66.36:37784 (5 connections now open)
2017-11-28T10:42:17.874+0000 I NETWORK  [conn39] end connection 192.168.66.36:37804 (4 connections now open)
2017-11-28T10:42:17.876+0000 I NETWORK  [conn35] end connection 192.168.66.36:37766 (3 connections now open)
2017-11-28T10:42:17.876+0000 I NETWORK  [conn38] end connection 192.168.66.36:37800 (2 connections now open)
2017-11-28T10:42:17.880+0000 I NETWORK  [conn36] SocketException handling request, closing client connection: 9001 socket exception [SEND_ERROR] server [192.168.66.36:37782]
2017-11-28T10:42:18.468+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37838 #41 (2 connections now open)
2017-11-28T10:42:18.493+0000 I NETWORK  [conn40] end connection 192.168.66.36:37818 (1 connection now open)
2017-11-28T10:42:30.084+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:37890 #42 (2 connections now open)
2017-11-28T11:59:57.381+0000 I NETWORK  [conn41] end connection 192.168.66.36:37838 (1 connection now open)
2017-11-28T11:59:57.388+0000 I NETWORK  [initandlisten] connection accepted from 192.168.66.36:48444 #43 (2 connections now open)
```

Error log from Cyclone Server:
```
[2017-11-28 18:32:04] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 18:32:38] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 18:32:38] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 18:32:38] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 18:32:38] [ERROR] [pipeline.go:179] Closed explicitly
E1128 18:41:55.186473       1 store.go:124] Ping Mongodb with error read tcp 192.168.66.36:56930->10.254.43.178:27017: i/o timeout
[2017-11-28 19:28:58] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:28:58] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:28:58] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:28:58] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:30:46] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:30:46] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:30:46] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:30:46] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:53:47] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:53:47] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:53:47] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:53:47] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:56:46] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:56:46] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:56:46] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:56:46] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:56:47] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:56:47] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 19:56:47] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 19:56:47] [ERROR] [pipeline.go:179] Closed explicitly
E1128 19:59:55.186168       1 store.go:124] Ping Mongodb with error read tcp 192.168.66.36:37838->10.254.43.178:27017: i/o timeout
[2017-11-28 20:00:24] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 20:00:24] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 20:00:24] [ERROR] [pipeline.go:160] Closed explicitly
[2017-11-28 20:00:24] [ERROR] [pipeline.go:179] Closed explicitly
[2017-11-28 20:00:29] [ERROR] [pipeline.go:160] Closed explicitly
```

--- a/store/store.go
+++ b/store/store.go
@@ -82,7 +82,9 @@ func Init(host string, gracePeriod time.Duration, closing chan struct{}) (*mgo.S
 	}
 
 	log.Infof("connect to mongodb addr: %s", host)
-	session.SetMode(mgo.Strong, true)
+	// Set the session mode as Eventual to ensure that the socket is created for each request.
+	// Can switch to other mode only after the old APIs are cleaned up.
+	session.SetMode(mgo.Eventual, true)
 
 	go backgroundMongo(closing)
 

